lib/types.ts

```diff
@@ -0,0 +1,85 @@
+import type * as typesObject from './constant/types'
+import type {ValueOf} from 'type-fest'
+
+export interface Point {
+  line: number
+  column: number
+  offset: number
+  _index: number
+  _bufferIndex: number
+}
+export interface Token {
+  start: Point
+  end: Point
+  type: unknown
+  contentType: string
+  previous: Token
+  next: Token
+  _size: number
+  _break?: boolean
+  _tokenizer?: unknown
+  _events?: Event[]
+  _subevents?: Event[]
+  _contentTokenized?: unknown
+  _close?: unknown
+  _marker?: unknown
+  _open?: unknown
+  _side?: number
+  _tabs?: unknown
+}
+
+export type Type = ValueOf<typeof typesObject>
+
+export type Event = [name: string, token: Token, context: unknown]
+
+export interface Effects {
+  enter: (type: Type) => Token
+  consume: (code: number) => Token
+  exit: (type: Type) => Token
+  attempt: (
+    one: unknown,
+    two: unknown,
+    three?: unknown
+  ) => (code: number) => unknown
+  interrupt: (
+    one: unknown,
+    two: unknown,
+    three: unknown
+  ) => (code: number) => unknown
+  check: (
+    one: unknown,
+    two: unknown,
+    three?: unknown
+  ) => (code: number) => unknown
+  lazy: (one: unknown, ok: Okay, nok: NotOkay) => void
```

Comment:
check, interrupt, and lazy are all “checks”: they are like attempt, but when succesful they *also* revert.
This is useful to see what lies ahead.

[interrupt](https://github.com/micromark/micromark/search?q=interrupt&unscoped_q=interrupt) is used for stuff right after a line of content.
[lazy](https://github.com/micromark/micromark/search?q=lazy&unscoped_q=lazy) is used for lines that were not properly preceded by the container.

interrupt and lazy work by setting a property on the parser (both their name). Tokenizers can then exit early if that is detected (e.g., when fenced code is started, it’s fenced code: no need to continue parsing more), or in cases where constructs cannot exist (setext heading underlines can’t be lazy)

The container parser (document) also sets these properties in certain cases directly.